Anima/diffusion/mcm_estimator/animaMCMEstimatorImageFilter.hxx

```diff
@@ -348,21 +348,74 @@ MCMEstimatorImageFilter<InputPixelType, OutputPixelType>
             m_MCMCreators[i]->SetConcentrationBounds(kappaLowerBound,kappaUpperBound);
     }
     
-    // Setting up coarse grids for kappa and nu initial search (useful for NODDI)
-    m_FractionCoarseGrid.resize(m_CoarseGridSize);
-    m_KappaCoarseGrid.resize(m_CoarseGridSize);
+    // Switch over compartment types to setup coarse grid initialization
+    switch (m_CompartmentType)
+    {
+        case NODDI:
+            this->ComputeNODDIExtraAxonalAndKappaCoarseGrids();
+            break;
+
+        case DDI:
+            itkExceptionMacro("DDI coarse grid not implemented yet");
+            break;
+
+        case Tensor:
+            this->ComputeTensorRadialDiffsAndAzimuthCoarseGrids();
+            break;
+
+        default:
+            // No coarse grid initialization for simple models
+            break;
+    }
+}
+
+template <class InputPixelType, class OutputPixelType>
+void
+MCMEstimatorImageFilter<InputPixelType, OutputPixelType>
+::ComputeNODDIExtraAxonalAndKappaCoarseGrids()
+{
+    // m_ValuesCoarseGrid[0] -> extra axonal fractions, m_ValuesCoarseGrid[1] -> kappa
+    m_ValuesCoarseGrid.resize(2);
+
+    unsigned int coarseGridSize = 8;
+    m_ValuesCoarseGrid[0].resize(coarseGridSize);
+    m_ValuesCoarseGrid[1].resize(coarseGridSize);
+
     // The anisotropy index suggested by NODDI signal expression is not the ODI
     // index defined in Zhang et al., 2012, Neuroimage. Instead, it amounts to
     // (3 * tau1(k) - 1) / 2 \in [0,1] which is not analytically invertible.
     // However it is well approximated by k^power / (halfLife + k^power) where
     // power and halfLife are given by:
+
     double power = 1.385829;
     double halfLife = 5.312364;
-    for (unsigned int i = 0;i < m_CoarseGridSize;++i)
+    for (unsigned int i = 0;i < coarseGridSize;++i)
+    {
+        double tmpVal = (i + 1.0) / (coarseGridSize + 1.0);
+        m_ValuesCoarseGrid[0][i] = tmpVal;
+        m_ValuesCoarseGrid[1][i] = std::pow(tmpVal * halfLife / (1.0 - tmpVal), 1.0 / power);
+    }
+}
+
+template <class InputPixelType, class OutputPixelType>
+void
+MCMEstimatorImageFilter<InputPixelType, OutputPixelType>
+::ComputeTensorRadialDiffsAndAzimuthCoarseGrids()
+{
+    // m_ValuesCoarseGrid[0] -> azimuth
+    // m_ValuesCoarseGrid[1] -> radial diff 1 vs axial diff weight
+    // m_ValuesCoarseGrid[2] -> radial diff 2 vs radial diff 1 weight
+    m_ValuesCoarseGrid.resize(3);
+
+    unsigned int coarseGridSize = 8;
+    for (unsigned int i = 0;i < 3;++i)
+        m_ValuesCoarseGrid[i].resize(coarseGridSize);
+
+    for (unsigned int i = 0;i < coarseGridSize;++i)
     {
-        double tmpVal = (i + 1.0) / (m_CoarseGridSize + 1.0);
-        m_FractionCoarseGrid[i] = tmpVal;
-        m_KappaCoarseGrid[i] = std::pow(tmpVal * halfLife / (1.0 - tmpVal), 1.0 / power);
+        m_ValuesCoarseGrid[0][i] = i * M_PI / (coarseGridSize - 1.0);
```

Comment:
The third angle in tensor parametrisation has an upper bound of 2pi not pi.
Also we should do as in NODDI initial grid and not include the bounds themselves (safer). I.e. ```(upperBound - lowerBound) * (i + 1.0) / (coarseGridSize + 1.0) + lowerBound```.